contracts/extensions/version-control/ConditionalImplementControl.sol

```diff
@@ -0,0 +1,164 @@
+    NEW_VERSION = newVersion_;
+    CURRENT_VERSION = currentVersion_;
+  }
+
+  /**
+   * @dev Fallback function that forwards the call to the current or new contract implementation based on a condition.
+   */
+  fallback() external payable virtual onlyDelegateFromProxyStorage {
+    _fallback();
+  }
+
+  receive() external payable virtual onlyDelegateFromProxyStorage {
+    _fallback();
+  }
+
+  /**
+   * @dev Executes the selfMigrate function, upgrading to the new contract implementation.
+   */
+  function selfMigrate() external onlyDelegateFromProxyStorage onlySelfCall {
+    _upgradeTo(NEW_VERSION);
+  }
+
+  /**
+   * @dev Internal function to get the current version of the contract implementation.
+   * @return The address of the current version.
+   */
+  function _getVersion() internal view virtual returns (address);
+
+  function _isConditionMet() internal view virtual returns (bool) {}
+
+  function _fallback() internal virtual {
+    bytes memory returnData = _dispatchCall(_getVersion());
+    assembly {
+      return(add(returnData, 0x20), mload(returnData))
+    }
+  }
+
+  /**
+   * @dev Internal function to dispatch the call to the specified version.
+   * @param version The address of the version to call.
+   * @return returnData The return data of the call.
+   */
+  function _dispatchCall(address version) internal virtual whenConditionsAreMet returns (bytes memory returnData) {
+    (bool success, bytes memory returnOrRevertData) = version.delegatecall(msg.data);
+    success.handleRevert(msg.sig, returnOrRevertData);
+    assembly {
+      returnData := returnOrRevertData
+    }
+  }
+
+  /**
+   * @dev Internal function to check if a contract address has code.
+   * @param addr The address of the contract to check.
+   * @dev Throws an error if the contract address has no code.
+   */
+  function _requireHasCode(address addr) internal view {
+    if (addr.code.length == 0) revert ErrZeroCodeContract(addr);
+  }
+
+  /**
+   * @dev Internal function to check if the caller is delegating from proxy storage.
+   * @dev Throws an error if the current implementation of the proxy storage is not this contract.
+   */
+  function _requireDelegateFromProxyStorage() private view {
+    if (address(this) != PROXY_STORAGE) revert ErrDelegateFromUnknownOrigin(address(this));
+  }
+
+  /**
+   * @dev Internal method to check method caller.
+   *
+   * Requirements:
+   *
+   * - The method caller must be this contract.
+   *
+   */
+  function _requireSelfCall() internal view virtual {
+    if (msg.sender != PROXY_STORAGE) revert ErrOnlySelfCall(msg.sig);
+  }
+
+  /**
+   * @dev Suggested gas stipend for contract to call selfMigrate function.
+   */
+  function _gasStipenedNoGrief() internal pure virtual returns (uint256) {
+    // Gas stipend for contract to perform a few
+    // storage reads and writes, but low enough to prevent griefing.
+    // Griefing means comsuming gas exhaustively when function call are reverted.
```

Comment:
```suggestion
    // Gas stipend for contract to perform a few read and write operations on storage, but
    // low enough to prevent comsuming gas exhaustively when function call are reverted.
```